component/metrics/scraper/appendable.go

```diff
@@ -0,0 +1,93 @@
+package scraper
+
+import (
+	"context"
+	"sync"
+
+	"github.com/prometheus/prometheus/model/value"
+
+	"github.com/prometheus/prometheus/model/labels"
+
+	"github.com/grafana/agent/component/metrics"
+	"github.com/prometheus/prometheus/model/exemplar"
+	"github.com/prometheus/prometheus/storage"
+)
+
+type scrapeAppendable struct {
+	mut sync.Mutex
+	// TODO add comment on why using a map when most of the time its only 1
+	buffer    map[int64][]*metrics.FlowMetric
+	receivers []*metrics.Receiver
+}
+
+func newScrapeAppendable(receiver []*metrics.Receiver) *scrapeAppendable {
+	return &scrapeAppendable{
+		buffer:    make(map[int64][]*metrics.FlowMetric),
+		receivers: receiver,
+	}
+}
+
+func (s *scrapeAppendable) Append(ref storage.SeriesRef, l labels.Labels, t int64, v float64) (storage.SeriesRef, error) {
+	s.mut.Lock()
+	defer s.mut.Unlock()
+
+	if len(s.receivers) == 0 {
+		return 0, nil
+	}
+	_, found := s.buffer[t]
+	if !found {
+		set := make([]*metrics.FlowMetric, 0)
+		s.buffer[t] = set
+	}
+	// If ref is 0 then lets grab a global id
+	if ref == 0 {
```

Comment:
@mattdurham Looking at the [Appendable interface](https://github.com/prometheus/prometheus/blob/f389711943daca716675fdd3b37901f4d70100f7/storage/interface.go#L204-L213) in Prometheus, the provided RefID is _optional_ and is used to accelerate calls. If zero is passed, then the matching happens through the labels, where either a new series is created, or the old refID is returned by matching the labels (code is [here](https://github.com/prometheus/prometheus/blob/f389711943daca716675fdd3b37901f4d70100f7/tsdb/head_append.go#L244-L301)).

The call to Append in the scrape loop along with the refID matching happens [here](https://github.com/prometheus/prometheus/blob/f389711943daca716675fdd3b37901f4d70100f7/scrape/scrape.go#L1552). As far as I can tell, Append will be called with a zero refID only if that series is not cached in the scrape loop, (and proceed with the label matching).

Here, we do things a little differently if I get it correctly; we depend on the seriesRef being zero to recognize that it's a new series, at which point we generate a new global refID.

I'm not certain how a series could be dropped from the scrape loop's cache, (a reload of an upstream component maybe?) but what's the worst thing that could happen in our usecase? Is it just that we'd be creating a new refID for a previously-seen series? Would that previous entry ultimately get GCed?

Reply:
No worries, it was a dumb name. Code that is hard to read is just as bad as bad code ;)